scalatest/src/main/scala/org/scalatest/prop/Randomizer.scala

```diff
@@ -280,47 +328,113 @@ class Randomizer(private[scalatest] val seed: Long) { thisRandomizer =>
+    // - (2 to the power of 23 (23 is the number of mantissa bits))
+    // + (2 for the two infinities)
+
+    // scala> Int.MaxValue - math.pow(2, 23).toInt + 2
+    // res13: Int = 2139095041
+    val (x, r) = chooseInt(Int.MinValue, 2139095041) 
+
+    // Pick two lucky numbers to play the lotto with:
+    if (x == 777)
+      (Float.NegativeInfinity, r)
+    else if (x == 888)
+      (Float.PositiveInfinity, r)
+    else r.nextFiniteFloatValue
   }
 
   /**
     * Get a random non-infinite Double.
     *
     * This can return either a positive or negative value, or zero, but guards against
-    * returning either [[Double.PositiveInfinity]] or [[Double.NegativeInfinity]].
+    * returning either [[Double.PositiveInfinity]], [[Double.NegativeInfinity]], or [[Double.NaN]].
     *
     * @return A random finite Double, and the next Randomizer to use.
     */
   def nextFiniteDouble: (FiniteDouble, Randomizer) = {
-    val (n, r) = nextDouble // TODO: Study nextFloat and nextDouble to see if it produces NaNs or Infinities.
-    val finite =            // See if it produces non-normal (less than max precision) values
-      n match {
-        case Double.PositiveInfinity => Double.MaxValue
-        case Double.NegativeInfinity => Double.MaxValue
-        case _ => n
-      }
-    (FiniteDouble.ensuringValid(finite), r)
+    // The exponent portion of a Double occupies 11 bits. It can be 0 (which represents an exponent of -1023)
+    // to 2047 (which is a reserved value for NaN values and +/- infinity). The highest regular (non-reserved)
+    // exponent therefore is 2046, which in hex is 0x7fe (which represents an exponent of +1023).
+    // Thus by chosing the exponent Int between 0 and 0x7fe, we can't get a NaN or an infinity.
+    val (s, rs) = chooseLong(0L, 1L)                  // The sign bit (1 bit)
+    val (e, re) = rs.chooseLong(0L, 0x7feL)           // The exponent (11 bits)
+    val (m, rm) = re.chooseLong(0L, 0xfffffffffffffL) // The mantissa (52 bits)
+    val finite = java.lang.Double.longBitsToDouble((s << 63) | (e << 52) | m)
+    (FiniteDouble.ensuringValid(finite), rm)
+  }
+
+  private def nextFiniteDoubleValue: (Double, Randomizer) = {
+    val (fd, r) = nextFiniteDouble
+    (fd.value, r)
+  }
+
+  // An extended real float includes all the numeric ones plus positive and negative infinity
+  // In other words, the whole space other than the NaNs. This may be a missing abstraction in anyvals.
+  // Let's think about that.
+  private def nextExtRealDoubleValue: (Double, Randomizer) = {
+
+    // The space of possible non-NaN Doubles
+    // 2 to the power of 64, which is covered by Long.MinValue to Long.MaxValue
+    // - (2 to the power of 52 (52 is the number of mantissa bits))
+    // + (2 for the two infinities)
+
+    // scala> Long.MaxValue - math.pow(2, 52).toLong + 2
+    // res27: Long = 9218868437227405313
+    val (x, r) = chooseLong(Long.MinValue, 9218868437227405313L) 
```

Comment:
Same comment as before: I'd write this in hex.

Reply:
Will do.